File: src/lib/api/templates.ts
/**
 * Execution template IPC wrappers.
 */
import { invoke } from "@tauri-apps/api/core";
import type { OrchestrateResult } from "./orchestrate";
import type { TaskNode, TaskRun } from "./tasks";

function isTauri(): boolean {
  return typeof window !== "undefined" && "__TAURI_INTERNALS__" in window;
}

export type TemplateVariable = {
  key: string;
  label: string;
  required: boolean;
  default?: string | null;
};

export type Template = {
  id: string;
  name: string;
  description: string | null;
  source_goal_id: string | null;
  source_plan_id: string | null;
  source_run_id: string | null;
  goal_prompt: string;
  plan_json: string;
  variables_json: string;
  created_at: string;
  updated_at: string;
};

export type PolishResult = {
  ok: boolean;
  name: string | null;
  description: string | null;
  goal_prompt: string | null;
  plan_json: string | null;
  variables: TemplateVariable[];
  raw_output: string | null;
  error: string | null;
};

export type PolishTemplateResult = {
  polish: PolishResult;
  source_goal_id: string | null;
  source_plan_id: string | null;
  source_run_id: string | null;
};

export type DispatchResult = {
  run: TaskRun;
  nodes: TaskNode[];
};

export type StartRunResult = {
  run_id: string;
  started: boolean;
};

export type InstantiateTemplateResult = {
  orchestrate: OrchestrateResult;
  dispatch: DispatchResult | null;
  started: StartRunResult | null;
};

export async function listTemplates(): Promise<Template[]> {
  if (!isTauri()) return [];
  return invoke<Template[]>("list_templates");
}

export async function getTemplate(id: string): Promise<Template | null> {
  if (!isTauri()) return null;
  return invoke<Template | null>("get_template", { id });
}

export async function polishTemplate(args: {
  goalId?: string | null;
  planId?: string | null;
  runId?: string | null;
  goalPrompt?: string | null;
  planJson?: string | null;
  skipAi?: boolean;
}): Promise<PolishTemplateResult> {
  if (!isTauri()) {
    throw new Error("polishTemplate requires Tauri runtime");
  }
  return invoke<PolishTemplateResult>("polish_template", {
    args: {
      goal_id: args.goalId ?? null,
      plan_id: args.planId ?? null,
      run_id: args.runId ?? null,
      goal_prompt: args.goalPrompt ?? null,
      plan_json: args.planJson ?? null,
      skip_ai: args.skipAi ?? false,
    },
  });
}

export async function createTemplate(args: {
  name: string;
  description?: string | null;
  sourceGoalId?: string | null;
  sourcePlanId?: string | null;
  sourceRunId?: string | null;
  goalPrompt: string;
  planJson: string;
  variablesJson?: string | null;
}): Promise<Template> {
  if (!isTauri()) {
    throw new Error("createTemplate requires Tauri runtime");
  }
  return invoke<Template>("create_template", {
    args: {
      name: args.name,
      description: args.description ?? null,
      source_goal_id: args.sourceGoalId ?? null,
      source_plan_id: args.sourcePlanId ?? null,
      source_run_id: args.sourceRunId ?? null,
      goal_prompt: args.goalPrompt,
      plan_json: args.planJson,
      variables_json: args.variablesJson ?? "[]",
    },
  });
}

export async function updateTemplate(args: {
  id: string;
  name?: string | null;
  description?: string | null;
  goalPrompt?: string | null;
  planJson?: string | null;
  variablesJson?: string | null;
}): Promise<Template> {
  if (!isTauri()) {
    throw new Error("updateTemplate requires Tauri runtime");
  }
  return invoke<Template>("update_template", {
    args: {
      id: args.id,
      name: args.name ?? null,
      description: args.description ?? null,
      goal_prompt: args.goalPrompt ?? null,
      plan_json: args.planJson ?? null,
      variables_json: args.variablesJson ?? null,
    },
  });
}

export async function deleteTemplate(id: string): Promise<void> {
  if (!isTauri()) {
    throw new Error("deleteTemplate requires Tauri runtime");
  }
  await invoke("delete_template", { id });
}

export async function duplicateTemplate(id: string): Promise<Template> {
  if (!isTauri()) {
    throw new Error("duplicateTemplate requires Tauri runtime");
  }
  return invoke<Template>("duplicate_template", { id });
}

export async function instantiateTemplate(args: {
  templateId: string;
  values: Record<string, string>;
  dispatch: boolean;
  concurrency?: number | null;
}): Promise<InstantiateTemplateResult> {
  if (!isTauri()) {
    throw new Error("instantiateTemplate requires Tauri runtime");
  }
  return invoke<InstantiateTemplateResult>("instantiate_template", {
    args: {
      template_id: args.templateId,
      values: args.values,
      dispatch: args.dispatch,
      concurrency: args.concurrency ?? null,
    },
  });
}

export function parseTemplateVariables(raw: string): TemplateVariable[] {
  try {
    const v = JSON.parse(raw) as TemplateVariable[];
    return Array.isArray(v) ? v : [];
  } catch {
    return [];
  }
}

File: src-tauri/src/commands/templates.rs
//! Template IPC — CRUD, polish, instantiate.
use crate::commands::orchestrate::{
    orchestrate_from_json, OrchestrateFromJsonArgs, OrchestrateResult,
};
use crate::commands::tasks::{start_run, StartRunResult};
use crate::repo::{
    create_template as repo_create, delete_template as repo_delete,
    duplicate_template as repo_duplicate, get_goal, get_orchestrator_settings, get_plan,
    get_task_run, get_template, list_templates as repo_list, update_template as repo_update,
    Template, TemplateCreate, TemplateUpdate,
};
use crate::services::dispatch::{dispatch_plan as svc_dispatch, DispatchResult};
use crate::services::orchestrate::{parse_plan_json, validate_plan};
use crate::services::template_polish::{
    polish_template_blocking, skip_polish_draft, PolishResult,
};
use crate::services::template_vars::{
    assert_plan_structure_unchanged, instantiate_texts, parse_variables_json, resolve_values,
    unknown_placeholders,
};
use crate::state::{DbState, RunState};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tauri::{AppHandle, State};

fn with_db<T, F>(state: &State<'_, DbState>, f: F) -> Result<T, String>
where
    F: FnOnce(&rusqlite::Connection) -> Result<T, String>,
{
    let conn = state
        .conn
        .lock()
        .map_err(|e| format!("db lock poisoned: {e}"))?;
    f(&conn)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PolishTemplateArgs {
    #[serde(default)]
    pub goal_id: Option<String>,
    #[serde(default)]
    pub plan_id: Option<String>,
    #[serde(default)]
    pub run_id: Option<String>,
    #[serde(default)]
    pub goal_prompt: Option<String>,
    #[serde(default)]
    pub plan_json: Option<String>,
    /// If true, skip CLI polish and return raw snapshot.
    #[serde(default)]
    pub skip_ai: bool,
}

fn load_goal_plan(
    conn: &rusqlite::Connection,
    args: &PolishTemplateArgs,
) -> Result<(String, String, Option<String>, Option<String>, Option<String>), String> {
    let mut source_goal_id = args.goal_id.clone();
    let mut source_plan_id = args.plan_id.clone();
    let mut source_run_id = args.run_id.clone();

    if let Some(run_id) = &args.run_id {
        let wrapped = get_task_run(conn, run_id)?
            .ok_or_else(|| format!("run not found: {run_id}"))?;
        source_goal_id = Some(wrapped.run.goal_id.clone());
        source_plan_id = Some(wrapped.run.plan_id.clone());
        source_run_id = Some(wrapped.run.id.clone());
    }

    if let (Some(g), Some(p)) = (
        args.goal_prompt.as_deref().map(str::trim).filter(|s| !s.is_empty()),
        args.plan_json.as_deref().map(str::trim).filter(|s| !s.is_empty()),
    ) {
        return Ok((
            g.to_string(),
            p.to_string(),
            source_goal_id,
            source_plan_id,
            source_run_id,
        ));
    }

    let plan_id = source_plan_id
        .clone()
        .ok_or_else(|| "plan_id or plan_json required".to_string())?;
    let plan = get_plan(conn, &plan_id)?.ok_or_else(|| format!("plan not found: {plan_id}"))?;
    let goal_id = source_goal_id.clone().unwrap_or_else(|| plan.goal_id.clone());
    let goal = get_goal(conn, &goal_id)?.ok_or_else(|| format!("goal not found: {goal_id}"))?;
    Ok((
        goal.prompt,
        plan.analysis_json,
        Some(goal_id),
        Some(plan_id),
        source_run_id,
    ))
}

#[tauri::command]
pub fn list_templates(state: State<'_, DbState>) -> Result<Vec<Template>, String> {
    with_db(&state, repo_list)
}

#[tauri::command(rename = "get_template")]
pub fn get_template_cmd(
    state: State<'_, DbState>,
    id: String,
) -> Result<Option<Template>, String> {
    with_db(&state, |c| get_template(c, &id))
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PolishTemplateResult {
    pub polish: PolishResult,
    pub source_goal_id: Option<String>,
    pub source_plan_id: Option<String>,
    pub source_run_id: Option<String>,
}

#[tauri::command]
pub async fn polish_template(
    state: State<'_, DbState>,
    args: PolishTemplateArgs,
) -> Result<PolishTemplateResult, String> {
    let (goal_prompt, plan_json, source_goal_id, source_plan_id, source_run_id) = {
        let conn = state
            .conn
            .lock()
            .map_err(|e| format!("db lock poisoned: {e}"))?;
        load_goal_plan(&conn, &args)?
    };

    // Validate plan parses
    let _ = parse_plan_json(&plan_json)?;

    if args.skip_ai {
        return Ok(PolishTemplateResult {
            polish: skip_polish_draft(&goal_prompt, &plan_json),
            source_goal_id,
            source_plan_id,
            source_run_id,
        });
    }

    let settings = {
        let conn = state
            .conn
            .lock()
            .map_err(|e| format!("db lock poisoned: {e}"))?;
        get_orchestrator_settings(&conn)?
    };

    let goal_prompt_c = goal_prompt.clone();
    let plan_json_c = plan_json.clone();
    let polish = tauri::async_runtime::spawn_blocking(move || {
        polish_template_blocking(&settings, &goal_prompt_c, &plan_json_c)
    })
    .await
    .map_err(|e| format!("polish join error: {e}"))?;

    Ok(PolishTemplateResult {
        polish,
        source_goal_id,
        source_plan_id,
        source_run_id,
    })
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateTemplateArgs {
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub source_goal_id: Option<String>,
    #[serde(default)]
    pub source_plan_id: Option<String>,
    #[serde(default)]
    pub source_run_id: Option<String>,
    pub goal_prompt: String,
    pub plan_json: String,
    #[serde(default)]
    pub variables_json: Option<String>,
}

#[tauri::command(rename = "create_template")]
pub fn create_template_cmd(
    state: State<'_, DbState>,
    args: CreateTemplateArgs,
) -> Result<Template, String> {
    let vars_raw = args.variables_json.unwrap_or_else(|| "[]".into());
    let vars = parse_variables_json(&vars_raw)?;
    let unknown = unknown_placeholders(&args.goal_prompt, &args.plan_json, &vars)?;
    if !unknown.is_empty() {
        return Err(format!(
            "undeclared placeholders: {}",
            unknown.join(", ")
        ));
    }
    // Parse plan for basic shape
    let _ = parse_plan_json(&args.plan_json)?;

    with_db(&state, |c| {
        repo_create(
            c,
            TemplateCreate {
                name: args.name,
                description: args.description,
                source_goal_id: args.source_goal_id,
                source_plan_id: args.source_plan_id,
                source_run_id: args.source_run_id,
                goal_prompt: args.goal_prompt,
                plan_json: args.plan_json,
                variables_json: vars_raw,
            },
        )
    })
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateTemplateArgs {
    pub id: String,
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub goal_prompt: Option<String>,
    #[serde(default)]
    pub plan_json: Option<String>,
    #[serde(default)]
    pub variables_json: Option<String>,
}

#[tauri::command(rename = "update_template")]
pub fn update_template_cmd(
    state: State<'_, DbState>,
    args: UpdateTemplateArgs,
) -> Result<Template, String> {
    with_db(&state, |c| {
        let existing = get_template(c, &args.id)?
            .ok_or_else(|| format!("template not found: {}", args.id))?;

        let plan_json = if let Some(ref new_plan) = args.plan_json {
            Some(assert_plan_structure_unchanged(
                &existing.plan_json,
                new_plan,
            )?)
        } else {
            None
        };

        let goal_prompt = args
            .goal_prompt
            .clone()
            .unwrap_or_else(|| existing.goal_prompt.clone());
        let final_plan = plan_json
            .clone()
            .unwrap_or_else(|| existing.plan_json.clone());
        let vars_raw = args
            .variables_json
            .clone()
            .unwrap_or_else(|| existing.variables_json.clone());
        let vars = parse_variables_json(&vars_raw)?;
        let unknown = unknown_placeholders(&goal_prompt, &final_plan, &vars)?;
        if !unknown.is_empty() {
            return Err(format!(
                "undeclared placeholders: {}",
                unknown.join(", ")
            ));
        }

        repo_update(
            c,
            &args.id,
            TemplateUpdate {
                name: args.name,
                description: args.description,
                goal_prompt: args.goal_prompt,
                plan_json,
                variables_json: args.variables_json,
            },
        )
    })
}

#[tauri::command(rename = "delete_template")]
pub fn delete_template_cmd(state: State<'_, DbState>, id: String) -> Result<(), String> {
    with_db(&state, |c| repo_delete(c, &id))
}

#[tauri::command(rename = "duplicate_template")]
pub fn duplicate_template_cmd(
    state: State<'_, DbState>,
    id: String,
) -> Result<Template, String> {
    with_db(&state, |c| repo_duplicate(c, &id))
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstantiateTemplateArgs {
    pub template_id: String,
    /// key → value
    #[serde(default)]
    pub values: HashMap<String, String>,
    /// If true, dispatch + start after creating plan.
    #[serde(default)]
    pub dispatch: bool,
    /// User override for this run only (1–8). Absent → use plan.concurrency / default.
    #[serde(default)]
    pub concurrency: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstantiateTemplateResult {
    pub orchestrate: OrchestrateResult,
    pub dispatch: Option<DispatchResult>,
    pub started: Option<StartRunResult>,
}

#[tauri::command]
pub fn instantiate_template(
    app: AppHandle,
    state: State<'_, DbState>,
    runs: State<'_, RunState>,
    args: InstantiateTemplateArgs,
) -> Result<InstantiateTemplateResult, String> {
    let (goal, plan_json, template_id) = {
        let conn = state
            .conn
            .lock()
            .map_err(|e| format!("db lock poisoned: {e}"))?;
        let tmpl = get_template(&conn, &args.template_id)?
            .ok_or_else(|| format!("template not found: {}", args.template_id))?;
        let declared = parse_variables_json(&tmpl.variables_json)?;
        let resolved = resolve_values(&declared, &args.values)?;
        // Also require values for any undeclared leftover — substitute will catch
        let (goal, plan_json) =
            instantiate_texts(&tmpl.goal_prompt, &tmpl.plan_json, &resolved)?;

        // Agent presence check before direct dispatch
        if args.dispatch {
            let plan = parse_plan_json(&plan_json)?;
            let validated = validate_plan(&conn, plan)?;
            let _ = validated;
        }

        (goal, plan_json, tmpl.id)
    };

    let orch = orchestrate_from_json(
        state.clone(),
        OrchestrateFromJsonArgs {
            goal,
            plan_json,
            template_key: Some(template_id),
        },
    )?;

    if !orch.ok {
        return Ok(InstantiateTemplateResult {
            orchestrate: orch,
            dispatch: None,
            started: None,
        });
    }

    if !args.dispatch {
        return Ok(InstantiateTemplateResult {
            orchestrate: orch,
            dispatch: None,
            started: None,
        });
    }

    let plan_id = orch
        .plan_row
        .as_ref()
        .map(|p| p.id.clone())
        .ok_or_else(|| "orchestrate succeeded but plan_row missing".to_string())?;

    let dispatched = with_db(&state, |c| svc_dispatch(c, &plan_id))?;
    let started = start_run(app, state, runs, dispatched.run.id.clone(), args.concurrency)?;

    Ok(InstantiateTemplateResult {
        orchestrate: orch,
        dispatch: Some(dispatched),
        started: Some(started),
    })
}
